### PR TITLE
Fix form submit behavior

### PR DESCRIPTION
**Pull request by shubham1206agra**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31433
PROPOSAL: https://github.com/Expensify/App/issues/31433#issuecomment-1814452374

### Tests

1. Open App
2. Tap profile icon
3. Tap profile > Personal Details > Legal Name.
4. Enter any text.
5. Tap enter.
6. Note the user is navigated to "personal details" page.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests

### QA Steps

Same as Tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/58412969/b9ddd2c3-7817-46f8-b2ff-eac51de33fc1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/58412969/8eba0096-65fd-4298-8ffa-ff5ad804b3da

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/58412969/eab78231-978d-43b9-9b7f-a079fad4d850

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/58412969/94daa8af-a98b-4fc9-b5e4-564f91beee4a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/58412969/6ff581a7-5353-4065-a643-afe61d4c377f

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/58412969/66f97d07-a21d-43b5-ab8e-b16ce2047c09

</details>

### CODE REVIEW

**Comment by shubham1206agra** [comment]
@cubuspl42 Should I modify the logic used here for native devices? https://github.com/Expensify/App/pull/24636

**Comment by shubham1206agra** [comment]
> Would you quickly sum up our two basic options here? What will be the consequences (especially the ones observable with the naked eye) of **modifying** the mentioned `submitOnEnter` handling on Native and **not modifying** it?

I think when the implementation was done, they haven't evaluated the option of native platform. Right now, it does not submit on enter in mWeb devices only (not on native). I think we should modify it for not submit on native devices as it will allow us to press enter for multi-line field in native.

**Comment by cubuspl42** [comment]
So, currently, we can't add a multi-line description on Native? Or did I get this wrong

**Comment by shubham1206agra** [comment]
> So, currently, we can't add a multi-line description on Native? Or did I get this wrong

No. Wait let me reach out on slack to clarify.

**Comment by shubham1206agra** [comment]
Yes I will ping you

**Comment by shubham1206agra** [comment]
Reproducible issue (also present on main) [NAB]
Submit does not work on 1st iteration, and on 2nd iteration, it goes 2 steps back after submit.

**Comment by cubuspl42** [comment]
> Reproducible issue (also present on main) [NAB] Submit does not work on 1st iteration, and on 2nd iteration, it goes 2 steps back after submit.

Do we have a tracking issue for this?

**Comment by shubham1206agra** [comment]
> > Reproducible issue (also present on main) [NAB] Submit does not work on 1st iteration, and on 2nd iteration, it goes 2 steps back after submit.
> 
> Do we have a tracking issue for this?

Nope

**Comment by shubham1206agra** [comment]
I am not sure why performance test is failing.

**Comment by cubuspl42** [comment]
I noticed that [you mentioned](https://github.com/Expensify/App/issues/31433#issuecomment-1900391352) we're holding on https://github.com/Expensify/App/issues/34474

**Comment by cubuspl42** [comment]
> I am not sure why performance test is failing.

Typically it means they are failing on `main` and merging it helps after some time passes. If not, it might mean we broke something performance-wise.

**Comment by shubham1206agra** [comment]
> I noticed that [you mentioned](https://github.com/Expensify/App/issues/31433#issuecomment-1900391352) we're holding on #34474

Actually I just opened this for review so that you can review this as I am OOO from 24th to 29th January.

File: src/components/RoomNameInput/index.native.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ import * as RoomNameInputUtils from '@libs/RoomNameInputUtils';
 import CONST from '@src/CONST';
 import * as roomNameInputPropTypes from './roomNameInputPropTypes';
 
-function RoomNameInput({isFocused, autoFocus, disabled, errorText, forwardedRef, value, onBlur, onChangeText, onInputChange, shouldDelayFocus}) {
+function RoomNameInput({isFocused, autoFocus, disabled, errorText, forwardedRef, value, onBlur, onChangeText, onInputChange, onSubmitEditing, returnKeyType, shouldDelayFocus}) {
```
Comment (cubuspl42) [inline src/components/RoomNameInput/index.native.js:10]: I don't understand this change. We haven't adjusted the use of this component, have we? Why is this change necessary?

File: src/pages/EditRequestDescriptionPage.js [hunk 2]
```diff
@@ -77,7 +77,7 @@ function EditRequestDescriptionPage({defaultDescription, onSubmit}) {
                         }}
                         autoGrowHeight
                         containerStyles={[styles.autoGrowHeightMultilineInput]}
-                        submitOnEnter={!Browser.isMobile()}
+                        submitOnEnter={!canUseTouchScreen()}
```
Comment (cubuspl42) [inline src/pages/EditRequestDescriptionPage.js:80]: I don't think that `submitOnEnter` is a good name, especially in the context of our recent changes

Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js:80]: We should also add a comment that this is a property intended only for effectively multiline input

Comment (cubuspl42) [inline src/pages/EditRequestDescriptionPage.js:80]: Could we walk over what's actually happening here and why?

We want to make this effectively multiline text input be able to submit the form ("submit on Enter"), but only when we can't use the touch screen (think: desktop web). What is the behavior in such a case? How do we add a new line? Has the behavior changed?

**Comment by cubuspl42** [comment]
Overall it looks good! 👍 I left some questions

File: src/pages/EditRequestDescriptionPage.js (hunk 2 above)
Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: In web where we have keyboard, we generally use Shift + enter to enter a new line. But in phones, we can't use shift + enter as it is not available. So we don't submit here at all on enter.

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: Can you suggest a new name which you seem fit here?
And I will add a comment indicating the same.

Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js:80]: It this case-by-case logic? Do we ever want a multi-line input that you can't reasonably make multiline on mobile?

I'm wondering whether this shouldn't just be handled centrally, with a nice code comment rephrasing what you just said.

Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js:80]: I think we can "hold" this thread on [this topic](https://github.com/Expensify/App/pull/34471#discussion_r1461807307), as we might just nuke this property.

File: src/components/Form/FormProvider.js [hunk 3]
```diff
@@ -260,6 +260,22 @@ const FormProvider = forwardRef(
...
+                const onSubmitEditingObject = shouldSubmitEdit
+                    ? {
+                          onSubmitEditing: (e) => {
+                              submit();
+                              if (!propsToParse.onSubmitEditing) {
+                                  return;
+                              }
+                              propsToParse.onSubmitEditing(e);
+                          },
+                      }
```
Comment (cubuspl42) [inline src/components/Form/FormProvider.js:274]: Would you add `returnKeyType` to this object?

`onSubmitEditingObject` could be renamed to `formSubmitObject`.

We handle this two things in an inconsistent manner.

Comment (cubuspl42) [inline src/components/Form/FormProvider.js:274]: Would you add `returnKeyType` to this object?

`onSubmitEditingObject` could be renamed to `formSubmitObject`.

We handle this two things in an inconsistent manner.

File: src/components/Form/FormProvider.js [hunk 4]
```diff
@@ -260,6 +260,22 @@ const FormProvider = forwardRef(
...
+                    : {};
+
+                const isMultiline = propsToParse.multiline || propsToParse.autoGrowHeight;
```
Comment (cubuspl42) [inline src/components/Form/FormProvider.js:277]: Doesn't this logic duplicate [this one](https://github.com/Expensify/App/pull/34471/files#diff-abde879b2542bb8feeb124550730f95fc100a648cc25683a5340271b4a7dc5adR20) in any way?

File: src/components/Form/FormProvider.js [hunk 5]
```diff
@@ -272,6 +288,9 @@ const FormProvider = forwardRef(
 
                 return {
                     ...propsToParse,
+                    returnKeyType: shouldSubmitEdit ? 'go' : propsToParse.returnKeyType,
+                    blurOnSubmit: (isMultiline && shouldSubmitEdit) || propsToParse.blurOnSubmit,
```
Comment (shubham1206agra) [inline src/components/Form/FormProvider.js:292]: Anwering https://github.com/Expensify/App/pull/34471/files#r1461816468

I have to pass blurOnSubmit on multiline input only. So thats why I am calculating isMultiline again.

File: src/components/Form/FormProvider.js (hunk 3 above)
Comment (shubham1206agra) [reply src/components/Form/FormProvider.js:274]: I will do that.

File: src/pages/EditRequestDescriptionPage.js (hunk 2 above)
Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: You have raised a good point.
We can do that. But I am not sure we are are allowed to do that.

Let me summon @luacmartins as form manager to decide here.

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: Not really. We are not allowing submit on enter on every multiline input even on keyboard devices (desktop). So we need to decide a name here.

I am thinking `forceSubmitOnEnter` would be better.

File: src/components/Form/FormProvider.js (hunk 5 above)
Comment (cubuspl42) [reply src/components/Form/FormProvider.js:292]: Let's also hold on [this thing](https://github.com/Expensify/App/pull/34471/files#r1461807307), as the surrounding code might change if we go that route

File: src/components/RoomNameInput/index.native.js (hunk 1 above)
Comment (shubham1206agra) [reply src/components/RoomNameInput/index.native.js:10]: We have changed the logic here as RoomNameInput has a TextInput to make it as close to the previous behavior.

File: src/components/Form/FormProvider.js (hunk 3 above)
Comment (shubham1206agra) [reply src/components/Form/FormProvider.js]: Done.

File: src/pages/EditRequestDescriptionPage.js (hunk 2 above)
Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js:80]: I'm having a problem with the "Enter" part. I know that it's a mess.

In my mental model, the submitting key ("Go" / "Done" / (arrow) ...) is presented _instead_ of the Enter/Return key).

If I recall correctly, pressing the submitting key on the virtual keyboard has a different effect from pressing the actual "Enter" key on the physical external Bluetooth keyboard connected to an iPhone or iPad. Possibly similar on Android devices.

This is complicated by the mobile Web browsers, which don't expose the native submit behavior to DOM API. This might be a backwards-compatibility leftover from times when all touch events could only be handled by DOM mouse events.

That's why I'm insisting on dropping the "Enter" part from the name and focusing on the model where a text input can be "submitted". On the Web (`react-native-web`) this is implemented as listening for the "Enter" key (which on native iOS/Android, again, _could_ be the press of the virtual submitting key) from a lack of a better option.

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: Ok should we call it `forceSubmitOnReturnKey` as RN also calls it a Return Key?

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: This logic has a flaw. It will be always be false for multi-line inputs.

Comment (luacmartins) [reply src/pages/EditRequestDescriptionPage.js:80]: What exactly are you trying to do? I see `whether this shouldn't just be handled centrally`, but what does that mean?

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js:80]: We have to decide between 2 options:
1. Developer will never be able to submit form through enter key on multiline input in mobile devices (both mWeb/native) by hardcoding the conditions inside FormProvider.
2. Let it left as it is on the developer choice (this is current implementation).

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js]: Thanks for the input.

Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js]: Not what I meant, and I don't think the latest change is reasonable 🙁

I meant this to be the _centralized_ logic we discussed before:

```
inputAllowsSubmit = single-line -> true; multi-line -> is_desktop_web
```

Just a local, in-component conditional computation. 

And _this_, as I stated explicitly, I meant to be the component property. This is the reincarnation of `submitOnEnter`, with an adjusted contract:

```
property shouldSubmitForm (default: for single-line -> true; for multi-line -> false)
```

Technically, it won't be a "React default", as it would be tricky technically. Technically, it will be `undefined` by default in the React terms, and then we'll calculate an effective default inside the component.

Try implementing this, or raise concerns if you disagree with it. If you are still unsure what I mean, I can implement the commit and let you cherry-pick it.

**Comment by luacmartins** [comment]
We have conflicts now

**Review by luacmartins — CHANGES_REQUESTED** [review]
We have conflicts now

File: src/pages/EditRequestDescriptionPage.js (hunk 2 above)
Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js]: @shubham1206agra Please let me know if I worded this clearly

Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js]: I may be misunderstanding something here. Can you give me commit diff, and then we can discuss this?

Comment (cubuspl42) [reply src/pages/EditRequestDescriptionPage.js]: Would you review it and test it? I could've missed something.

**Comment by luacmartins** [comment]
@shubham1206agra lint is failing

File: src/pages/EditRequestDescriptionPage.js (hunk 2 above)
Comment (shubham1206agra) [reply src/pages/EditRequestDescriptionPage.js]: Your logic is mostly fine with some minor bugs, which I am fixing.

**Comment by shubham1206agra** [comment]
@cubuspl42 Can you give this another round of review?

File: src/components/Form/InputWrapper.tsx [hunk 6]
```diff
@@ -18,11 +18,13 @@ function computeComponentSpecificRegistrationParams<TInput extends ValidInputs>(
...
     if (InputComponent === TextInput) {
-        const isEffectivelyMultiline = Boolean(multiline ?? autoGrowHeight);
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        const isEffectivelyMultiline = Boolean(multiline || autoGrowHeight);
 
         // We calculate the effective requested value of `shouldSubmitForm`, assuming that the default value should be
         // `true` for single-line inputs and `false` for multi-line inputs.
-        const shouldSubmitFormOrDefault = shouldSubmitForm ?? !isEffectivelyMultiline;
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        const shouldSubmitFormOrDefault = shouldSubmitForm || !isEffectivelyMultiline;
```
Comment (cubuspl42) [inline src/components/Form/InputWrapper.tsx:30]: Why is this a fix? What is your intention here? As I see it, it makes it impossible to request `shouldSubmitForm = false` for single line inputs.

File: src/components/Form/InputWrapper.tsx [hunk 7]
```diff
@@ -18,11 +18,13 @@ function computeComponentSpecificRegistrationParams<TInput extends ValidInputs>(
     readonly shouldSetTouchedOnBlurOnly: boolean,
 } {
     if (InputComponent === TextInput) {
-        const isEffectivelyMultiline = Boolean(multiline ?? autoGrowHeight);
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        const isEffectivelyMultiline = Boolean(multiline || autoGrowHeight);
```
Comment (cubuspl42) [inline src/components/Form/InputWrapper.tsx:25]: You're right; it was not correct. What do you think about...

```ts
Boolean(multiline) || Boolean(autoGrowHeight)
```

This code is explicit, as it acknowledges that these values aren't proper booleans (but `boolean | undefined`), and it needs no linter silencing I think.

File: src/components/Form/InputWrapper.tsx (hunk 6 above)
Comment (shubham1206agra) [reply src/components/Form/InputWrapper.tsx:30]: And that should be the case for single line input as per `FORMS.md`.

Comment (cubuspl42) [reply src/components/Form/InputWrapper.tsx:29]: Hmm... So maybe we could nuke `shouldSubmitFormOrDefault` and `canInputSubmitForm`, and go with a single expression in such a case.

```ts
const shouldReallySubmitForm = isEffectivelyMultiline ? Boolean(shouldSubmitForm) && canUseHardwareKeyboard ? true;
```

`shouldReallySubmitForm` could get this comment:

> We want to avoid a situation when the user can't insert a new line. For single-line inputs, it's not a problem and we force-enable form submission. For multi-line inputs, ensure that it was requested to enable form submission for this specific input and that alternative ways exist to add a new line.

My intentions were good, but I might have made it unnecessarily complex.

File: src/components/Form/InputWrapper.tsx [hunk 8]
```diff
@@ -1,21 +1,74 @@
...
+    multiline,
+    autoGrowHeight,
+    blurOnSubmit,
+}: InputWrapperProps<TInput>): {
+    readonly shouldSubmitForm: boolean;
+    readonly blurOnSubmit: boolean | undefined;
+    readonly shouldSetTouchedOnBlurOnly: boolean;
+} {
+    const validTextInputComponents = [TextInput, AddressSearch, RoomNameInput] as TInput[];
+    if (validTextInputComponents.includes(InputComponent)) {
```
Comment (cubuspl42) [inline src/components/Form/InputWrapper.tsx:23]: Nit: a new line before `if`?

**Review by jakub-trzebiatowski — CHANGES_REQUESTED** [review]

File: src/components/Form/InputWrapper.tsx (hunk 6 above)
Comment (shubham1206agra) [reply src/components/Form/InputWrapper.tsx]: Done.

**Comment by cubuspl42** [comment]
Please test "Personal Details > Address > Address Line 1" on Android/Web. 

See what the submitting button behavior was before (on `main`) and what is after the changes. It doesn't seem to work now.

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/fc8d4815-e6d7-4d8d-a3d9-f8224bb0795c

Seems to be working for me. Can you share a video for the same?

**Comment by shubham1206agra** [comment]
https://github.com/Expensify/App/assets/58412969/214b7391-1cae-4a8b-9ed6-a496eef03d6e

Still working :)

**Comment by cubuspl42** [comment]
Thanks for checking this! It didn't work for me; I'm not sure why. I'll re-test.

**Comment by cubuspl42** [comment]
Please test a case, where you input a single word, like "Somewhere", to "Address Line 1". Don't accept any suggestions, but try to submit the input with the manual entry. This works everywhere except Android/Web for me.

Also, I'm having doubts about whether, in this specific case (address input), the new behavior is better. Address search is not a typical text input; it was never clear whether this is actually in the scope of this issue/PR.

Please repeat the test I suggested. If you reproduce this issue on Android/Web, I suggest removing the address search out of the PR scope instead of debugging and fixing it.

**Comment by cubuspl42** [comment]
Also, start with merging `main`

**Comment by shubham1206agra** [comment]
@cubuspl42 I am not able to repro the issue. Can you provide me with the video? It will really help me to repro the issue.

https://github.com/Expensify/App/assets/58412969/12896361-6695-49da-bd25-4372cfd7ce35

**Comment by cubuspl42** [comment]
> I am not able to repro the issue.

When I test on **Android/Web**, I observe this result:

https://github.com/Expensify/App/assets/2590174/492f2eec-7701-425a-ae33-3a0d1e3a1ad7

**Comment by luacmartins** [comment]
@shubham1206agra did you get a chance to look at @cubuspl42's [last comment](https://github.com/Expensify/App/pull/34471#issuecomment-1926635386)?

**Comment by luacmartins** [comment]
Bump @shubham1206agra. We also have conflicts now. Let's try to get this one to the finish line!

**Comment by shubham1206agra** [comment]
> > I am not able to repro the issue.
> 
> When I test on **Android/Web**, I observe this result:
> 
>  form-submit-issue-1-android-web-compressed.mp4 
> This seems to be specific to Android/Web and single-word locations.

@cubuspl42 I still am not able to repro this. Let me ask an adhoc build here.
cc @luacmartins

File: src/components/Form/InputWrapper.tsx [hunk 9]
```diff
@@ -1,21 +1,66 @@
...
+    shouldSubmitForm,
+    multiline,
+    autoGrowHeight,
+    blurOnSubmit,
+}: InputWrapperProps<TInput>): {
+    readonly shouldSubmitForm: boolean;
+    readonly blurOnSubmit: boolean | undefined;
+    readonly shouldSetTouchedOnBlurOnly: boolean;
+} {
+    const textInputBasedComponents = [TextInput, RoomNameInput] as TInput[];
```
Comment (cubuspl42) [inline src/components/Form/InputWrapper.tsx:21]: Type assertions should be avoided at a high cost. They are a very dangerous type system feature.

As I understand this, this typing will also work:

```ts
const textInputBasedComponents: ValidInputs[] = [TextInput, RoomNameInput];
```

**Review by jakub-trzebiatowski — CHANGES_REQUESTED** [review]

File: src/components/Form/InputWrapper.tsx (hunk 9 above)
Comment (shubham1206agra) [reply src/components/Form/InputWrapper.tsx]: Thanks for the suggestion. I was unable to use this correctly at first.

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/2590174/70116d01-be05-4c18-8361-4a1660552eb7

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/2590174/6c3c3ca3-ed36-4e5c-980a-c4a7a81591d2

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/2590174/eb41ca2e-5171-4bd7-aa0a-b79e9e7e99ae

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/2590174/6723fcb9-b0b3-49fe-803f-df65bee2c109

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/2590174/88277f05-f62c-4b46-8c34-1d85bf7e49d8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/2590174/64e730ae-da20-4921-877c-cadbcb184eef

</details>

**Comment by cubuspl42** [comment]
I'm still having problems on Android/Web, this time also with plain text inputs. Maybe this is something on my side?...

@luacmartins Would you have a possibility to test this PR on Android/Web?

**Comment by luacmartins** [comment]
@cubuspl42 sorry, I'm only working 50% today and won't have time to do that today. It'd be great if you could figure this out. Otherwise I can try testing tomorrow.

**Comment by cubuspl42** [comment]
I'll try testing on an old physical Android smartphone I own. Let me charge it.

**Review by jakub-trzebiatowski — APPROVED** [review]
It's working on my physical Android device on Chrome 👍

**Review by luacmartins — APPROVED** [review]
LGTM!

**Comment by hayata-suenaga** [comment]
this issue might be related to [this deploy blocker](https://github.com/Expensify/App/issues/36370)
